### PR TITLE
Allow non-read-only external libraries that support asset deletion

### PR DESCRIPTION
**Pull request by valpackett**
*Related: https://github.com/immich-app/immich/discussions/5449 https://github.com/immich-app/immich/discussions/8438*
*Sponsored by: @benmccann*

This is an attempt at switching from hardcoded "external means cannot delete" to an explicit `isReadOnly` flag for libraries. (Not fully polished change yet, hence draft)

The per-asset `isExternal` and `isReadOnly` flags being stored in the database is making this difficult… I guess just using them and updating them while the library's read-onliness changes would be significantly easier than removing them and making everywhere query the library flags, but I'm looking for feedback re: the right way to do this.

### CODE REVIEW

**Comment by alextran1502** [comment]
I think the most natural change of this feature is to let the user know they can mount the library with `ro` or without that flag for docker mount; then, we can treat external library assets as normal upload assets to avoid introducing an additional boolean flag in the system. What do you think?

**Comment by alextran1502** [comment]
@etnoy I am pretty sure `isReadOnly` flag isn't used anymore after we move to your implementation of external library, correct? It was for the previous import mechanism

**Comment by alextran1502** [comment]
> Hm. That would mean checking mount info (only on boot? on rescan too?) and updating the flags based on that. Seems nice in terms of not adding any admin UI. I'm slightly worried about potential surprising behavior (e.g. adding a writable mount, possibly for multiple libraries inside of it, and adding libraries in the admin interface, not expecting them to be automatically writable) but on the other hand "it's a purely outside configured sysadmin level decision" does feel nice in a way.
> 
> Not sure we'd actually avoid the flag on the library, even if it would get set automatically based on the mount: it would definitely make sense to show in the admin interface whether it is writable, and have that fully consistent with what we've set for the assets…

I am thinking more of, if the user know that they use the `ro` flag in docker mount, the deletion call will error out expectedly, otherwise, the operation will happen as if it the assets are from the upload library. So we don't necessary need to do anything special here in term of checking the state of read/write of the mount

**Comment by etnoy** [comment]
> @etnoy I am pretty sure `isReadOnly` flag isn't used anymore after we move to your implementation of external library, correct? It was for the previous import mechanism

That is my understanding, too.

I've argued for removing the field but I think @jrasm91 knew of something that still used it

**Comment by jrasm91** [comment]
I would recommend keeping the per asset  flags. Otherwise it will probably be a nightmare trying to load in the property everywhere it needs to be used. If a library is marked as read only, it seems easy enough to do a bulk asset update

File: server/src/dtos/library.dto.ts [hunk 1]
```diff
@@ -128,6 +131,8 @@ export class LibraryResponseDto {
   createdAt!: Date;
   updatedAt!: Date;
   refreshedAt!: Date | null;
+
+  isReadOnly!: boolean | null;
```
Comment (jrasm91) [inline server/src/dtos/library.dto.ts:135]: This isn't nullable.

File: server/src/entities/library.entity.ts [hunk 2]
```diff
@@ -53,6 +53,9 @@ export class LibraryEntity {
 
   @Column({ type: 'boolean', default: true })
   isVisible!: boolean;
+
+  @Column({ type: 'boolean', nullable: true })
+  isReadOnly!: boolean | null;
```
Comment (jrasm91) [inline server/src/entities/library.entity.ts:58]: This should not be nullable and should default to true

File: server/src/services/asset.service.ts [hunk 3]
```diff
@@ -372,8 +372,10 @@ export class AssetService {
       return JobStatus.FAILED;
     }
 
-    // Ignore requests that are not from external library job but is for an external asset
-    if (!fromExternal && (!asset.library || asset.library.type === LibraryType.EXTERNAL)) {
+    // Ignore requests that are not from external library job but is for an external read-only asset
+    const isReadOnlyLibrary = !asset.library || asset.library.isReadOnly ||
+      (asset.library.isReadOnly === null && asset.library.type === LibraryType.EXTERNAL);
+    if (!fromExternal && isReadOnlyLibrary) {
```
Comment (jrasm91) [inline server/src/services/asset.service.ts:378]: This should just check if from external and the asset is read only and skip if so. No reason to make it complicated.

File: server/src/services/asset.service.ts [hunk 4]
```diff
@@ -406,7 +408,7 @@ export class AssetService {
     }
 
     const files = [asset.thumbnailPath, asset.previewPath, asset.encodedVideoPath];
-    if (!(asset.isExternal || asset.isReadOnly)) {
+    if (!isReadOnlyLibrary) {
```
Comment (jrasm91) [inline server/src/services/asset.service.ts:411]: If asset isn't readonly

File: web/src/lib/components/asset-viewer/asset-viewer-nav-bar.svelte [hunk 5]
```diff
@@ -193,7 +193,7 @@
     {/if}
 
     {#if isOwner}
-      {#if !asset.isReadOnly || !asset.isExternal}
+      {#if !asset.isReadOnly || !asset.isExternal || (true /* XXX */)}
```
Comment (jrasm91) [inline web/src/lib/components/asset-viewer/asset-viewer-nav-bar.svelte:196]: Just change this to asset not read only

File: web/src/routes/admin/library-management/+page.svelte [hunk 6]
```diff
@@ -243,6 +243,19 @@
     updateLibraryIndex = selectedLibraryIndex;
   };
 
+  const onToggleReadOnlyClicked = async () => {
+    const library = libraries[selectedLibraryIndex];
+    const prevValue = library.isReadOnly == null ? (library.type === LibraryType.External) : library.isReadOnly;
+
+    try {
+      await updateLibrary({ id: library.id, updateLibraryDto: { ...library, isReadOnly: !prevValue } });
```
Comment (jrasm91) [inline web/src/routes/admin/library-management/+page.svelte:251]: Please don't send the whole dto. Just send the read only setting.

File: web/src/routes/admin/library-management/+page.svelte [hunk 7]
```diff
@@ -397,6 +410,12 @@
                     <Portal target="body">
                       <ContextMenu {...contextMenuPosition} on:outclick={() => onMenuExit()}>
                         <MenuOption on:click={() => onRenameClicked()} text={`Rename`} />
+                        <MenuOption
+                          on:click={() => onToggleReadOnlyClicked()}
+                          text={(selectedLibrary?.isReadOnly ||
```
Comment (jrasm91) [inline web/src/routes/admin/library-management/+page.svelte:415]: This should not be nullable

**Review by jrasm91 — COMMENTED** [review]
This seems fine, but should be simplified. Furthermore, when a library is updated and readonly changes we should SQL update the affected assets.

File: web/src/lib/components/asset-viewer/asset-viewer-nav-bar.svelte (hunk 5 above)
Comment (benmccann) [inline web/src/lib/components/asset-viewer/asset-viewer-nav-bar.svelte:196]:
```suggestion
      {#if !asset.isReadOnly}
```

File: web/src/routes/admin/library-management/+page.svelte (hunk 7 above)
Comment (benmccann) [reply web/src/routes/admin/library-management/+page.svelte:415]: we could probably change the `{#if showContextMenu}` above to `{#if showContextMenu && selectedLibrary}` to simplify all the accesses to `selectedLibrary` within

File: open-api/immich-openapi-specs.json [hunk 8]
```diff
@@ -8778,6 +8778,10 @@
             },
             "type": "array"
           },
+          "isReadOnly": {
+            "nullable": true,
```
Comment (benmccann) [inline open-api/immich-openapi-specs.json:8782]: I think this shouldn't be nullable either

**Comment by jrasm91** [comment]
Hey, thanks for working on this and sorry for the confusion/back and forth around this topic. Internally we have had a lot of discussions about the direction we want to take external libraries. Long story short, we decided to go ahead with the implementation laid out in #9235.

**Pull request closed without merging**